FiltersEngine.cpp

```diff
@@ -0,0 +1,201 @@
+*/
+#include "FiltersEngine.h"
+
+#include "Logger.h"
+
+#include <string>
+#include <iostream>
+#include <fstream>
+#include <sys/stat.h> /* for stat() */
+#include <dirent.h>
+#include <algorithm>
+#include <unistd.h>
+#include <limits.h>
+
+std::bitset<FILTER_BITSET_SIZE> FiltersEngine::AddFilter(ProcFilterSpec& pfs, std::string& outputName)
+{
+    std::bitset<FILTER_BITSET_SIZE> ret;
+
+    if (_filtersBitPosition.count(pfs) > 0) {
+        struct FiltersInfo info = _filtersBitPosition[pfs];
+        ret[info.bitPosition] = 1;
+        if (info.outputs.count(outputName) == 0) {
+            info.outputs.emplace(outputName);
+            _filtersBitPosition[pfs] = info;
+        }
+    } else {
+        struct FiltersInfo info;
+        info.bitPosition = _nextBitPosition;
+        info.outputs.emplace(outputName);
+        _filtersBitPosition[pfs] = info;
+        ret[_nextBitPosition] = 1;
+        _bitPositionSyscalls[_nextBitPosition] = pfs._syscalls;
+        _nextBitPosition++;
+    }
+
+    if (_outputs.count(outputName) == 0) {
+        _outputs.insert(outputName);
+    }
+
+    return ret;
+}
+
+std::bitset<FILTER_BITSET_SIZE> FiltersEngine::AddFilterList(std::vector<ProcFilterSpec>& pfsVec, std::string& outputName)
+{
+    std::bitset<FILTER_BITSET_SIZE> ret;
+
+    for (auto pfs : pfsVec) {
+        ret |= AddFilter(pfs, outputName);
+    }
+
+    return ret;
+}
+
+bool FiltersEngine::ProcessMatchFilter(std::shared_ptr<ProcessTreeItem> process, ProcFilterSpec& pfs, unsigned int height)
+{
+    if (pfs._depth != -1 && pfs._depth < height) {
+        return false;
+    }
+
+    if (pfs._match_mask & PFS_MATCH_UID) {
+        if (pfs._uid != process->_uid) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_GID) {
+        if (pfs._gid != process->_gid) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_EXE_EQUALS) {
+        if (pfs._exeMatchValue != process->_exe) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_EXE_STARTSWITH) {
+        if (process->_exe.substr(0, pfs._exeMatchValue.length()) != pfs._exeMatchValue) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_EXE_CONTAINS) {
+        if (process->_exe.find(pfs._exeMatchValue) == std::string::npos) {
+            return false;
+        }
+    }
+    if (pfs._match_mask & PFS_MATCH_EXE_REGEX) {
+        if (!std::regex_search(process->_exe, pfs._exeRegex)) {
+            return false;
+        }
+    }
+
+    for (auto cf : pfs._cmdlineFilters) {
+        if (cf._matchType == MatchEquals) {
+            if (cf._matchValue != process->_cmdline) {
+                return false;
+            }
+        } else if (cf._matchType == MatchStartsWith) {
+            if (process->_cmdline.substr(0, cf._matchValue.length()) != cf._matchValue) {
```

Comment:
Same here, please consider using starts_with() in StringsUtils.

Reply:
Accepted.